lib/Request.php

```diff
@@ -117,6 +120,13 @@ function setAbsoluteUrl(string $url) {
      */
     function getAbsoluteUrl() : string {
 
+        if (!$this->absoluteUrl) {
+            // Guessing we're a http endpoint.
+            $this->absoluteUrl = 'http://' .
```

Comment:
So we kind of do both already. Usually the absolute url comes from the Sapi. It actually populates `absoluteUrl` using the information from `$_SERVER` and tends to do this accurately.

In cases where you are using the object in a client, you generally populate the information yourself. A request without an absolute url cannot be made really. This is really the ultimate fallback if all else fails.